iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
 
-                if (TryStop())
+                if (result.ReasonCode != MqttClientPublishReasonCode.Success)
                 {
-                    OnTransportClosedGracefully();
-
-                    await _closeRetryPolicy.RunWithRetryAsync(CleanUpImplAsync, cancellationToken).ConfigureAwait(true);
+                    throw new IotHubClientException($"Failed to publish the mqtt packet for patching this client's twin with reason code {result.ReasonCode}", true);
                 }
-                else if (State == TransportState.Error)
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, $"Sent twin patch request with request id {requestId}. Now waiting for the service response.");
+
+                // Wait until IoT hub sends a message to this client with the response to this patch twin request.
+                PatchTwinResponse patchTwinResponse =
+                    await GetTaskCompletionSourceResultAsync(
+                        taskCompletionSource,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+
+                if (Logging.IsEnabled)
+                    Logging.Info(this, $"Received twin patch response for request id {requestId} with status {patchTwinResponse.Status}.");
+
+                if (patchTwinResponse.Status != 204)
                 {
-                    _fatalException.Throw();
+                    throw new IotHubClientException(patchTwinResponse.Message);
                 }
+
+                //TODO new twin version should be returned here, but API surface doesn't currently allow it
+                //return patchTwinResponse.Version;
+            }
+            catch (Exception e) when (e is not IotHubClientException && e is not OperationCanceledException)
+            {
+                throw new IotHubClientException("Failed to send twin patch", e, true);
             }
             finally
             {
-                if (Logging.IsEnabled)
-                    Logging.Exit(this, "", $"{nameof(MqttTransportHandler)}.{nameof(CloseAsync)}");
+                // No matter what, remove the requestId from this dictionary since no thread will be waiting for it anymore
+                _reportedPropertyUpdateResponseCompletions.TryRemove(requestId, out TaskCompletionSource<PatchTwinResponse> _);
             }
         }
 
-        #endregion Client operations
-
-        #region MQTT callbacks
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            _mqttClient?.Dispose();
+        }
 
-        public void OnConnected()
+        public override async Task CloseAsync(CancellationToken cancellationToken)
         {
-            if (TryStateTransition(TransportState.Opening, TransportState.Open))
+            OnTransportClosedGracefully();
+            MqttClientDisconnectOptions disconnectOptions = new MqttClientDisconnectOptions();
```

Comment:
Consider using `var` instead of `MqttClientDisconnectOptions`

Reply:
Also, inline it with the method call.